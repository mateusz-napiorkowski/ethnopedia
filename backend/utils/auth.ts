import { Request } from "express"
const jwt = require("jsonwebtoken")

export const checkUserIsLoggedIn = async (req: Request) => {
    try {
        const token = req.headers.authorization?.split(" ")[1]
        if (!token) return false;
        try {
            const decoded = jwt.verify(token, process.env.ACCESS_TOKEN_SECRET as string)
            return true
        } catch (error) {
            return false
        }
    } catch (error) {
        return false
    }
}

module.exports = { checkUserIsLoggedIn }