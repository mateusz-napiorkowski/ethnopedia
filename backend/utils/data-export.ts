import { subcategoryData } from "./interfaces"

const findValue: any = (subcategories: Array<subcategoryData>, categoryNameSplitByDot: Array<string>) => {
    const categoryDepth = categoryNameSplitByDot.length
    const [topmostParentCategoryName, ...categoryNameWithoutTopmostPart] = categoryNameSplitByDot;
    if(categoryDepth > 1) {
        const matchingSubcategory = subcategories.find(subcategory => subcategory.name === topmostParentCategoryName);
        return matchingSubcategory ? findValue(matchingSubcategory.subcategories, categoryNameWithoutTopmostPart) : '';
    }
    const matchingCategory = subcategories.find(category => category.name === topmostParentCategoryName);
    return matchingCategory ? matchingCategory.values.join(';') : '';
}

export const fillRow = (keys: Array<string>, categories: Array<subcategoryData>) => {
    const rowdata: any = {}
    keys.forEach(key => {
        rowdata[key] = findValue(categories, key.split("."))
    });
    return rowdata
}