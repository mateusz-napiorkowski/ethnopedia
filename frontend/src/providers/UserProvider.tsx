import React, { createContext, useContext, useState, useEffect } from "react"
import { jwtDecode } from "jwt-decode"

export interface JWT {
    exp: number;
    username: string;
    firstName: string;
    userId: string;
}

interface UserContextProps {
    isUserLoggedIn: boolean;
    firstName: string;
    userId: string;
    jwtToken: string;
    setUserData: (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => void;
}

export const UserContext = createContext<UserContextProps>({
    isUserLoggedIn: false,
    firstName: "",
    userId: "",
    jwtToken: "",
    setUserData: () => {
    },
})

export const useUser = () => useContext(UserContext)

interface UserProviderProps {
    children: React.ReactNode;
}

export const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
    const [isUserLoggedIn, setIsUserLoggedIn] = useState(false)
    const [firstName, setFirstName] = useState("")
    const [userId, setUserId] = useState("")
    const [jwtToken, setJwtToken] = useState("")

    const setUserData = (isLoggedIn: boolean, name: string, jwtToken: string, userId: string) => {
        setIsUserLoggedIn(isLoggedIn)
        setFirstName(name)
        setUserId(userId)
        setJwtToken(jwtToken)
    }

    useEffect(() => {
        const token = localStorage.getItem("token")

        if (token) {
            try {
                const decodedToken = jwtDecode<JWT>(token)
                const currentTime = Date.now() / 1000
                if (decodedToken.exp > currentTime) {
                    setUserData(true, decodedToken.firstName, token, decodedToken.userId)
                } else {
                    localStorage.removeItem("token")
                    setUserData(false, "", "", "")
                }
            } catch (error) {
                console.error("Invalid token:", error)
                localStorage.removeItem("token")
                setUserData(false, "", "", "")
            }
        }
    }, [])

    return (
        <UserContext.Provider value={{ isUserLoggedIn, firstName, userId, jwtToken, setUserData }}>
            {children}
        </UserContext.Provider>
    )
}
