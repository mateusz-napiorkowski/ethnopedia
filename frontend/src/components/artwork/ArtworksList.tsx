import { useQuery, useQueryClient } from "react-query"
import { useBatchDeleteArtworkMutation } from "../../api/artworks"
import { getArtworksInCollection } from "../../api/collections"
import LoadingPage from "../../pages/LoadingPage"
import React, { useEffect, useMemo, useState} from "react"
import Navbar from "../navbar/Navbar"
import { useLocation, useNavigate, useParams } from "react-router-dom"
import SearchComponent from "../search/SearchComponent"
import FileDropzone from "../FileDropzone"
import ExportOptions from "../ExportOptions"
import { ReactComponent as PlusIcon } from "../../assets/icons/plus.svg"
import { ReactComponent as FileImportIcon } from "../../assets/icons/fileImport.svg"
import { ReactComponent as FileExportIcon } from "../../assets/icons/fileExport.svg"
import WarningPopup from "../../pages/collections/WarningPopup"
import SortOptions from "../SortOptions"
import { getCollection, useBatchDeleteCollectionMutation } from "../../api/collections"
import Navigation from "../Navigation"
import Pagination from "../Pagination"
import { useUser } from "../../providers/UserProvider"

const ArtworksList = () => {
    const [selectedArtworks, setSelectedArtworks] = useState<{ [key: string]: boolean }>({})
    const [showFileDropzone, setShowFileDropzone] = useState<boolean>(false)
    const [showExportOptions, setShowExportOptions] = useState<boolean>(false)
    const [showDeleteRecordsWarning, setShowDeleteRecordsWarning] = useState(false)
    const [showDeleteCollectionWarning, setShowDeleteCollectionWarning] = useState(false)
    const [sortOrder, setSortOrder] = useState<string>("newest-first")
    const [showEditCollection, setShowEditCollection] = useState<boolean>(false)
    const { jwtToken } = useUser();
    const location = useLocation()
    useEffect(() => {

    }, [location]);

    const queryParameters = new URLSearchParams(window.location.search)
    const searchText = queryParameters.get("searchText")
    const [currentPage, setCurrentPage] = useState(1)
    const pageSize = 10

    const { collection } = useParams()
    const queryClient = useQueryClient()
    const { mutate: batchDeleteMutation } = useBatchDeleteArtworkMutation()

    const findValue = (artwork: any, categoryName: string) => {
        let val = ""
        artwork.categories.forEach((category: any) => {
            if(category.name === categoryName) {
                val = category.values.toString()
                return
            }
        });
        return val
    }

    const sortOptions = [
        { value: "newest-first", label: "Od najnowszych" },
        { value: "oldest-first", label: "Od najstarszych" },
        { value: "title-asc", label: "Tytuł rosnąco" },
        { value: "title-desc", label: "Tytuł malejąco" },
    ]

    const { data: artworkData} = useQuery({
        queryKey: ["artwork", currentPage, searchText, queryParameters, location, sortOrder],
        queryFn: () => getArtworksInCollection(collection as string, currentPage, pageSize, sortOrder, searchText, Object.fromEntries(queryParameters.entries())),
        enabled: !!collection,
    })

    const { data: collectionData } = useQuery({
        queryKey: [`${collection}`],
        enabled: !!collection,
        queryFn: () => getCollection(collection as string),
    })

    const selectAll = () => {
        const newSelection = artworkData?.artworks.reduce((acc: any, artwork: any) => {
            acc[artwork._id] = true
            return acc
        }, {})
        setSelectedArtworks(newSelection)
    }

    const sortArtworks = (artworks: any[], order: string) => {
        return artworks
    }

    const sortedArtworks = useMemo(() => {
        return sortArtworks([...(artworkData?.artworks || [])], sortOrder)
    }, [artworkData, sortOrder])

    const handleCheck = (id: string) => {
        setSelectedArtworks((prev) => ({ ...prev, [id]: !prev[id] }))
    }

    const deselectAll = () => {
        setSelectedArtworks({})
    }

    const deleteSelected = () => {
        const selectedIds = Object.keys(selectedArtworks).filter(id => selectedArtworks[id])

        if (selectedIds.length > 0) {
            batchDeleteMutation([selectedIds, jwtToken],
                {
                    onSuccess: () => {
                        queryClient.invalidateQueries(["artwork"])
                        setShowDeleteRecordsWarning(!showDeleteRecordsWarning)
                    },
                })
        }
    }

    const deleteCollectionMutation = useBatchDeleteCollectionMutation()
    const deleteCollection = () => {
        if (collection) {
            deleteCollectionMutation.mutate([collection], {
                onSuccess: () => {
                    queryClient.invalidateQueries(["artwork"])
                    queryClient.invalidateQueries(["collection"])
                    setShowDeleteCollectionWarning(false)
                    navigate("/")
                },
                onError: (error: any) => {
                    console.error("Error deleting collection: ", error)
                },
            })
        }
    }

    const navigate = useNavigate()

    if (artworkData === undefined || sortedArtworks === undefined) {
        return <LoadingPage />
    } else {
        const allArtworks = sortedArtworks.map((artwork: any) => (
            <div className="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
                 key={artwork._id}
                 onClick={() => navigate(`/collections/${collection}/artworks/${artwork._id}`)}>

                <div className="flex flex-row">
                        <span className="mr-4 flex items-center">
                            <input type="checkbox"
                                   checked={selectedArtworks[artwork._id!] || false}
                                   onClick={(e) => e.stopPropagation()}
                                   onChange={() => {
                                       handleCheck(artwork._id!)
                                   }} />
                        </span>

                    <div>
                        <h3 className="text-lg font-semibold text-gray-800 dark:text-white">{findValue(artwork, "Tytuł")}</h3>
                        <p className="text-gray-600 dark:text-gray-400 mb-1">{findValue(artwork, "Artyści")}</p>
                        <p className=" text-gray-500 dark:text-gray-300">{findValue(artwork, "Rok")}</p>
                    </div>
                </div>
            </div>
        ))

        return <>
            <Navbar />

            {/* {showFileDropzone && <FileDropzone onClose={() => setShowFileDropzone(false)} />}
            {showExportOptions && <ExportOptions onClose={() => setShowExportOptions(false)} />} */}
            {/*{showCreateArtwork && <CreateArtwork onClose={() => setShowCreateArtwork(false)} />}*/}
            {showDeleteRecordsWarning &&
                <WarningPopup onClose={() => setShowDeleteRecordsWarning(false)}
                              deleteSelected={deleteSelected}
                              warningMessage={"Czy na pewno chcesz usunąć zaznaczone rekordy?"} />}

            {showDeleteCollectionWarning &&
                <WarningPopup onClose={() => setShowDeleteCollectionWarning(false)}
                              deleteSelected={deleteCollection}
                              warningMessage={"Czy na pewno chcesz usunąć zaznaczoną kolekcję?"} />}
            <div className="flex flex-col w-full items-center bg-gray-50 dark:bg-gray-900 p-2 sm:p-4">
                <div className="flex flex-col max-w-screen-xl w-full lg:px-6">
                    <Navigation />

                    <div className="flex flex-row mb-4 mt-2">
                        <div className="flex flex-col w-full">
                            <h2 className="text-4xl font-bold text-gray-800 dark:text-white mb-1">
                                {collectionData?.name}
                            </h2>
                            <p className="text-xl text-gray-600 dark:text-gray-300">
                                {collectionData?.description}
                            </p>
                        </div>

                        {/* <div className="flex items-center">
                            <button className="text-lg mr-2 h-fit font-semibold ml-4"
                                    onClick={() => setShowEditCollection(true)}>
                                Edytuj
                            </button>
                        </div>

                        <div className="flex items-center">
                            <button
                                className="text-lg font-semibold h-fit border-red-700 text-red-700 bg-red-50 hover:bg-white">
                                <span className="flex-row flex items-center"
                                      onClick={() => setShowDeleteCollectionWarning(true)}>
                                    <p>Usuń</p>
                                </span>
                            </button>
                        </div> */}

                    </div>

                    {collection && <SearchComponent collectionName={collection} />}

                    <div className="flex w-full md:w-auto">
                        <div className="flex flex-1 space-x-2">
                            {jwtToken && <button className="flex items-center justify-center dark:text-white
                            hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                            dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                                 type="button"
                                                 onClick={() => navigate(`/collections/${collection}/create-artwork`)}>
                                    <span className="mr-2 text-white dark:text-gray-400">
                                        <PlusIcon />
                                    </span>
                                Nowy rekord
                            </button>}
                            {!jwtToken && <button disabled={true} title={"Aby dodać rekord musisz się zalogować."} className="flex items-center justify-center dark:text-white
                            hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                            dark:focus:ring-primary-800 font-semibold text-white hover:bg-gray-600 bg-gray-600 border-gray-800"
                                                  type="button"
                                                  onClick={() => navigate(`/collections/${collection}/create-artwork`)}>
                                    <span className="mr-2 text-white dark:text-gray-400">
                                        <PlusIcon />
                                    </span>
                                Nowy rekord
                            </button>}
                            <button className="flex items-center justify-center dark:text-white
                                            hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 font-medium px-4 py-2
                                            dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                    type="button"
                                    onClick={async () => {
                                        setShowExportOptions(showExportOptions => !showExportOptions)
                                    }}
                            >
                                <span className="text-white dark:text-gray-400">
                                    <FileExportIcon />
                                </span>
                                Eksportuj plik
                            </button>
                            {jwtToken && <button className="flex items-center justify-center dark:text-white
                                        hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                                        dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                                 type="button"
                                                 onClick={() => setShowFileDropzone(showFileDropzone => !showFileDropzone)}
                            >
                                <span className="text-white dark:text-gray-400">
                                    <FileImportIcon />
                                </span>
                                Importuj plik
                            </button>}
                            {!jwtToken && <button disabled={true} title={"Aby zaimportować plik musisz się zalogować."} className="flex items-center justify-center dark:text-white
                                        hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                                        dark:focus:ring-primary-800 font-semibold text-white bg-gray-600 hover:bg-gray-600 border-gray-800"
                                                  type="button"
                                                  onClick={() => setShowFileDropzone(showFileDropzone => !showFileDropzone)}
                            >
                                <span className="text-white dark:text-gray-400">
                                    <FileImportIcon />
                                </span>
                                Importuj plik
                            </button>}

                            <button className="flex items-center justify-center dark:text-white
                                        hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                                        dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                    type="button"
                                    onClick={selectAll}
                            >
                                Zaznacz wszystkie
                            </button>

                            <button className="flex items-center justify-center dark:text-white
                                        hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                                        dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                    type="button"
                                    onClick={deselectAll}
                            >
                                Odznacz wszystkie
                            </button>

                            {jwtToken && <button className="flex items-center justify-center dark:text-white
                                        hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                                        dark:focus:ring-primary-800 font-semibold text-white bg-gray-800 hover:bg-gray-700 border-gray-800"
                                                 type="button"
                                                 onClick={() => {
                                                     if (Object.keys(selectedArtworks).length !== 0)
                                                         setShowDeleteRecordsWarning(!showDeleteRecordsWarning)
                                                 }}
                            >
                                Usuń zaznaczone
                            </button>}
                            {!jwtToken && <button disabled={true} title={"Aby usuwać rekordy musisz się zalogować."} className="flex items-center justify-center dark:text-white
                                        hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 px-4 py-2
                                        dark:focus:ring-primary-800 font-semibold text-white bg-gray-600 hover:bg-gray-600 border-gray-800"
                                                  type="button"
                                                  onClick={() => {
                                                      if (Object.keys(selectedArtworks).length !== 0)
                                                          setShowDeleteRecordsWarning(!showDeleteRecordsWarning)
                                                  }}
                            >
                                Usuń zaznaczone
                            </button>}
                        </div>

                        <span className="">
                                <SortOptions
                                    options={sortOptions}
                                    onSelect={(value) => setSortOrder(value)}
                                />
                            </span>
                    </div>
                    {showFileDropzone && <FileDropzone onClose={() => setShowFileDropzone(false)} inCollectionPage={false}/>}
                    {showExportOptions && <ExportOptions selectedArtworks={selectedArtworks} onClose={() => setShowExportOptions(false)} />}
                </div>
            </div>

            <div className="flex flex-row">
                <div
                    className="flex mx-auto flex-1 justify-end w-full">
                    {/* <FilterDropdown /> */}
                </div>
                <div className="w-full flex-2 lg:px-6 max-w-screen-xl">
                    {allArtworks}
                </div>
                <div className="mx-auto w-full flex-1">
                </div>
            </div>
            <div className="flex justify-center mb-2">
                <Pagination
                    currentPage={currentPage}
                    totalPages={Math.ceil(artworkData.total / pageSize)}
                    setCurrentPage={(page) => setCurrentPage(page)}
                />
            </div>
        </>
    }
}
export default ArtworksList