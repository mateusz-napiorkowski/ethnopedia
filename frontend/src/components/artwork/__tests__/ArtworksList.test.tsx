import '@testing-library/jest-dom';
import { render, waitFor } from '@testing-library/react'
import ArtworksList from "../ArtworksList"
import { QueryClient, QueryClientProvider } from "react-query";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { UserContext } from '../../../providers/UserProvider';
import {collection, exampleArtworkId, artworksData, artworksDataSecondPage, collectionData, loggedInUserContextProps, jwtToken} from './utils/ArtworksListUtils'

const mockUseNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
   ...jest.requireActual('react-router-dom') as any,
  useNavigate: () => mockUseNavigate,
}));

const mockGetArtworksForCollectionPage = jest.fn()
const mockDeleteArtworks = jest.fn()
jest.mock('../../../api/artworks', () => ({
    getArtworksForCollectionPage: () => mockGetArtworksForCollectionPage,
    deleteArtworks: (artworkId: string, jwtToken: string) => mockDeleteArtworks(artworkId, jwtToken)
}))

const mockGetCollection = jest.fn()
jest.mock('../../../api/collections', () => ({
    getCollection: () => mockGetCollection
}))

const mockGetAllCategories = jest.fn()
jest.mock('../../../api/categories', () => ({
    getAllCategories: () => mockGetAllCategories
}))

const queryClient = new QueryClient();
const user = userEvent.setup()

const renderPage = (
    queryClient: QueryClient, 
    userContextProps: any = {
        isUserLoggedIn: false,
        firstName: "",
        userId: "",
        jwtToken: undefined,
        setUserData: jest.fn()
    },
    collection: string = "example collection",
    pageSize = 3
    ) => {
        return render(
            <UserContext.Provider value={ userContextProps }>
                <QueryClientProvider client={queryClient}>
                    <MemoryRouter initialEntries={[`/collections/${collection}/artworks/`]}>
                        <Routes>
                            <Route path="/collections/:collection/artworks/" element={<ArtworksList pageSize={pageSize}/>}/>
                        </Routes>  
                    </MemoryRouter>
                </QueryClientProvider>    
            </UserContext.Provider>
        );
};

describe("ArtworksList tests", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        queryClient.clear();
    });

    it("should render loading state", () => {     
        const {queryByTestId} = renderPage(queryClient)
        
        expect(queryByTestId("loading-page-container")).toBeInTheDocument()
        expect(queryByTestId("loaded-artwork-page-container")).not.toBeInTheDocument()
    })

    it("should render artworks list component after data is fetched from API, collection name and description should be displayed", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByTestId, queryByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))

        expect(queryByTestId("loading-page-container")).not.toBeInTheDocument()
        expect(getByTestId("loaded-artwork-page-container")).toBeInTheDocument()

        await waitFor(() => expect(getByTestId('collection-name-and-description-container')).toMatchSnapshot())
    })

    it("should have add record, import file, and delete selected buttons disabled when user is not logged in", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const addNewRecordButton = getByRole("button", {name: /nowy rekord/i})
        const importFileButton = getByRole("button", {name: /importuj plik/i})
        const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})

        expect(addNewRecordButton).toBeDisabled()
        expect(importFileButton).toBeDisabled()
        expect(deleteSelectedButton).toBeDisabled()
    })

    it("should have add record, import file, and delete selected buttons enabled when user is logged in", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const addNewRecordButton = getByRole("button", {name: /nowy rekord/i})
        const importFileButton = getByRole("button", {name: /importuj plik/i})
        const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})

        expect(addNewRecordButton).not.toBeDisabled()
        expect(importFileButton).not.toBeDisabled()
        expect(deleteSelectedButton).not.toBeDisabled()
    })
 
    it("should navigate to add record page when enabled add record button is clicked", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const addNewRecordButton = getByRole("button", {name: /nowy rekord/i})
        await user.click(addNewRecordButton)

        expect(mockUseNavigate).toHaveBeenCalledWith("/collections/example collection/create-artwork")
    })

    it("should open upload file window when enabled import file button is clicked", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByRole, getByTestId, getByText} = renderPage(queryClient, loggedInUserContextProps)

        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const importFileButton = getByRole("button", {name: /importuj plik/i})
        await user.click(importFileButton)

        expect(getByText(/prześlij plik/i)).toBeInTheDocument()
    })

    it("should open export options window when export file button is clicked", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        mockGetAllCategories.mockReturnValue({categories: ["Tytuł"]})
        const {getByRole, getByTestId, getByText, container} = renderPage(queryClient, loggedInUserContextProps)
        await waitFor(() => getByTestId('loaded-artwork-page-container'))
        const exportFileButton = getByRole("button", {name: /eksportuj plik/i})
        
        await user.click(exportFileButton)
        await waitFor(() => getByTestId("export-options-container"))
        
        expect(getByText(/ustawienia eksportu metadanych do pliku .xlsx/i)).toBeInTheDocument()
    })

    it("should navigate to appropriate artwork page when entry from artworks list is clicked", async () => {
        mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
        mockGetCollection.mockReturnValue(collectionData)
        const {getByTestId} = renderPage(queryClient)

        await waitFor(() => getByTestId('artworks-listed'))

        await user.click(getByTestId(exampleArtworkId))
        
        expect(mockUseNavigate).toHaveBeenCalledWith(`/collections/${collection}/artworks/${exampleArtworkId}`)
    })

    //TODO test disabled button when no artworks are selected
    // it("should not open delete selected warning popup when enabled delete selected button is clicked and no records are selected", async () => {
    //     mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
    //     mockGetCollection.mockReturnValue(collectionData)
    //     const {getByRole, getByTestId, queryByText} = renderPage(queryClient, loggedInUserContextProps)

    //     await waitFor(() => getByTestId('loaded-artwork-page-container'))
    //     const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})
    //     await user.click(deleteSelectedButton)

    //     expect(queryByText(/czy na pewno chcesz usunąć zaznaczone rekordy?/i)).not.toBeInTheDocument()
    // })

    //TODO page change tests
    // it("should refetch artworks list when artwork page is changed", async () => {
    //     mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
    //     mockGetCollection.mockReturnValue(collectionData)
    //     const {getByTestId} = renderPage(queryClient)

    //     await waitFor(() => getByTestId('loaded-artwork-page-container'))

    //     await waitFor(() => expect(getByTestId('artworks-listed')).toMatchSnapshot("page 1"))
    //     mockGetArtworksForCollectionPage.mockReturnValueOnce(artworksDataSecondPage)
    //     await user.click(getByTestId("page-2"))
    //     await waitFor(() => expect(getByTestId('artworks-listed')).toMatchSnapshot("page 2"))
    // })

    // TODO deleting of checked artworks tests
    // it("should ???", async () => {
    //     mockGetArtworksForCollectionPage.mockReturnValue(artworksData)
    //     mockGetCollection.mockReturnValue(collectionData)
    //     const {getByTestId, getByRole, getByLabelText, queryByText} = renderPage(queryClient, loggedInUserContextProps)

    //     await waitFor(() => getByTestId('artworks-listed'))

    //     await user.click(getByTestId(`${exampleArtworkId}-checkbox`))
        
    //     expect(getByTestId(`${exampleArtworkId}-checkbox`)).toBeChecked()

    //     const deleteSelectedButton = getByRole("button", {name: /usuń zaznaczone/i})
    //     await user.click(deleteSelectedButton)

    //     expect(queryByText("Czy na pewno chcesz usunąć zaznaczone rekordy?")).toBeInTheDocument()

    //     await user.click(getByLabelText("confirm"))

    //     await waitFor(() => expect(mockDeleteArtworks).toHaveBeenCalledWith([exampleArtworkId], jwtToken))
    //     expect(queryByText("Czy na pewno chcesz usunąć zaznaczone rekordy?")).not.toBeInTheDocument()
    // })
})