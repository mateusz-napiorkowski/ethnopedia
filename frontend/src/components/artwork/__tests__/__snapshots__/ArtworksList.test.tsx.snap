// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtworksList tests should refetch artworks list when artwork page is changed: page 1 1`] = `
<div
  class="w-full flex-2 lg:px-6 max-w-screen-xl"
  data-testid="artworks-listed"
>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752ddca46e3ca48231024dc"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752ddca46e3ca48231024dc-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <h3
          class="text-lg font-semibold text-gray-800 dark:text-white"
        >
          Jakym jechoł koło dworu
        </h3>
        <p
          class="text-gray-600 dark:text-gray-400 mb-1"
        >
          Józefa Piaskowska
        </p>
        <p
          class=" text-gray-500 dark:text-gray-300"
        >
          1960
        </p>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752ddca46e3ca48231024aa"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752ddca46e3ca48231024aa-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <h3
          class="text-lg font-semibold text-gray-800 dark:text-white"
        >
          Ojcze, ojcze, kup mi kónia
        </h3>
        <p
          class="text-gray-600 dark:text-gray-400 mb-1"
        >
          Magdalena Figlak
        </p>
        <p
          class=" text-gray-500 dark:text-gray-300"
        >
          1949
        </p>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752ddca46e3ca48231024bb"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752ddca46e3ca48231024bb-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <h3
          class="text-lg font-semibold text-gray-800 dark:text-white"
        >
          Piękna jagoda
        </h3>
        <p
          class="text-gray-600 dark:text-gray-400 mb-1"
        >
          Zespół Mazowsze
        </p>
        <p
          class=" text-gray-500 dark:text-gray-300"
        >
          2005
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ArtworksList tests should refetch artworks list when artwork page is changed: page 2 1`] = `
<div
  class="w-full flex-2 lg:px-6 max-w-screen-xl"
  data-testid="artworks-listed"
>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752aaaa46e3ca48231024dc"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752aaaa46e3ca48231024dc-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <h3
          class="text-lg font-semibold text-gray-800 dark:text-white"
        >
          W górach, nad rzeką
        </h3>
        <p
          class="text-gray-600 dark:text-gray-400 mb-1"
        >
          Józefa Piaskowska
        </p>
        <p
          class=" text-gray-500 dark:text-gray-300"
        >
          1960
        </p>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="6752aaaa46e5db48231024dc"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="6752aaaa46e5db48231024dc-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <h3
          class="text-lg font-semibold text-gray-800 dark:text-white"
        >
          Skrzypce na polu
        </h3>
        <p
          class="text-gray-600 dark:text-gray-400 mb-1"
        >
          Magdalena Figlak
        </p>
        <p
          class=" text-gray-500 dark:text-gray-300"
        >
          1949
        </p>
      </div>
    </div>
  </div>
  <div
    class="px-4 max-w-screen-xl py-4 bg-white dark:bg-gray-800 shadow-md w-full rounded-lg mb-4
                border border-gray-300 dark:border-gray-600 cursor-pointer"
    data-testid="1234aaca46e3ca48231024bb"
  >
    <div
      class="flex flex-row"
    >
      <span
        class="mr-4 flex items-center"
      >
        <input
          data-testid="1234aaca46e3ca48231024bb-checkbox"
          type="checkbox"
        />
      </span>
      <div>
        <h3
          class="text-lg font-semibold text-gray-800 dark:text-white"
        >
          Pragnom uoczka, pragnom
        </h3>
        <p
          class="text-gray-600 dark:text-gray-400 mb-1"
        >
          Zespół Mazowsze
        </p>
        <p
          class=" text-gray-500 dark:text-gray-300"
        >
          2005
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`ArtworksList tests should render artworks list component after data is fetched from API, collection name and description should be displayed 1`] = `
<div
  class="flex flex-row mb-4 mt-2"
  data-testid="collection-name-and-description-container"
>
  <div
    class="flex flex-col w-full"
  >
    <h2
      class="text-4xl font-bold text-gray-800 dark:text-white mb-1"
    >
      example collection
    </h2>
    <p
      class="text-xl text-gray-600 dark:text-gray-300"
    >
      example collection description
    </p>
  </div>
</div>
`;
